extensions/wikia/EmbeddableDiscussions/EmbeddableDiscussionsController.class.php

```diff
@@ -16,142 +16,90 @@ class EmbeddableDiscussionsController {
 
 	private $templateEngine;
 
-	function __construct() {
+	public function __construct() {
 		$this->templateEngine = ( new Wikia\Template\MustacheEngine )->setPrefix( __DIR__ . '/templates' );
 	}
 
-	public static function onParserFirstCallInit( Parser $parser ) {
-		global $wgEnableDiscussions;
+	public function render( array $args ) {
+		global $wgCityId;
 
-		if ( $wgEnableDiscussions ) {
-			$parser->setHook( self::TAG_NAME, [
-				'EmbeddableDiscussionsController',
-				'render'
-			] );
-		}
+		$parameters = [
+			static::PARAM_MOSTRECENT => $args[ static::PARAM_MOSTRECENT ] ?? false,
+			static::PARAM_SIZE => $args[ static::PARAM_SIZE ] ?? self::ITEMS_DEFAULT,
+			static::PARAM_COLUMNS => $args[ static::PARAM_COLUMNS ] ?? self::COLUMNS_DEFAULT,
+			static::PARAM_CATEGORY => $args[ static::PARAM_CATEGORY ] ?? '',
+		];
 
-		return true;
-	}
+		$errorMessage = $this->validateArguments( $parameters );
 
-	public static function onBeforePageDisplay() {
-		\Wikia::addAssetsToOutput( 'embeddable_discussions_js' );
-		\Wikia::addAssetsToOutput( 'embeddable_discussions_scss' );
+		$modelData = ( new DiscussionsThreadModel( $wgCityId ) )
+			->getData( $parameters[ static::PARAM_MOSTRECENT ], $parameters[ static::PARAM_SIZE ], $parameters[ static::PARAM_CATEGORY ] );
 
-		JSMessages::enqueuePackage( 'EmbeddableDiscussions', JSMessages::EXTERNAL );
+		$errorMessage = $this->validateCategory( $args, $modelData ) ?? $errorMessage;
 
-		return true;
+		if ( isset( $errorMessage ) ) {
+			return $this->templateEngine
+				->clearData()
+				->setData( [ 'errorMessage' => $errorMessage ] )
+				->render( 'DiscussionError.mustache' );
+		}
+
+		return F::app()->checkSkin( 'wikiamobile' ) ?
+			$this->renderMobile( $modelData, $parameters[ static::PARAM_MOSTRECENT ], $parameters[ static::PARAM_SIZE ] ) :
+			$this->renderDesktop( $modelData, $parameters[ static::PARAM_MOSTRECENT ], $parameters[ static::PARAM_CATEGORY ], $parameters[ static::PARAM_COLUMNS ] );
 	}
 
 	/**
-	 * Checks arguments for errors.
-	 * @param $modelData
-	 * @param array $args
-	 * @internal param $errorMessage /Return parameter with the proper error message to show.
-	 * Disregard if return is false.
-	 * @return true if ok, false if error
+	 * @param array $parameters
+	 * @return String error message or empty string if category is valid
 	 */
-	private function checkArguments( $modelData, array $args, &$errorMessage ) {
+	private function validateArguments( array $parameters ) {
+		$errorMessage = '';
+
 		// PARAM_MOSTRECENT must be bool
-		if ( isset( $args[ static::PARAM_MOSTRECENT ] ) &&
-			$args[ static::PARAM_MOSTRECENT ] !== 'true' &&
-			$args[ static::PARAM_MOSTRECENT ] !== 'false'
-		) {
+		if ( !AttributesValidator::isBoolish( $parameters[ static::PARAM_MOSTRECENT ] ) ) {
 			$errorMessage = wfMessage( 'embeddable-discussions-parameter-error',
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "embeddable-discussions-parameter-error" 4 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1192)